pkg/registry/core/pod/strategy.go

```diff
@@ -668,6 +669,56 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
 	return container, nil
 }
 
+// applyLabelKeysToLabelSelector obtains the label value from the given label set by the key in labelKeys,
+// and merge to LabelSelector with the given operator:
+func applyLabelKeysToLabelSelector(labelSelector *metav1.LabelSelector, labelKeys []string, operator metav1.LabelSelectorOperator, podLabels map[string]string) {
+	for _, key := range labelKeys {
+		if value, ok := podLabels[key]; ok {
+			labelSelector.MatchExpressions = append(labelSelector.MatchExpressions, metav1.LabelSelectorRequirement{
+				Key:      key,
+				Operator: operator,
+				Values:   []string{value},
+			})
+		}
+	}
+}
+
+// applyMatchLabelKeysAndMismatchLabelKeys obtains the labels from the pod labels by the key in matchLabelKeys or mismatchLabelKeys,
+// and merge to LabelSelector of PodAffinityTerm depending on field:
+// - If matchLabelKeys, key in (value) is merged with LabelSelector.
+// - If mismatchLabelKeys, key notin (value) is merged with LabelSelector.
+func applyMatchLabelKeysAndMismatchLabelKeys(term *api.PodAffinityTerm, label map[string]string) {
+	if (len(term.MatchLabelKeys) == 0 && len(term.MismatchLabelKeys) == 0) || term.LabelSelector == nil {
+		// If LabelSelector is nil, we don't need to apply label keys to it because nil-LabelSelector is match none.
+		return
+	}
+
+	applyLabelKeysToLabelSelector(term.LabelSelector, term.MatchLabelKeys, metav1.LabelSelectorOpIn, label)
```

Comment:
how does the scheduler react if both term.LabelSelector.MatchLabels and term.LabelSelector.MatchExpressions are set?

Reply:
They should be correctly handled in such cases as well.

1. `labels.Selector` is generated from LabelSelector of affinity (metav1.LabelSelector). [`LabelSelectorAsSelector` func](https://github.com/kubernetes/kubernetes/blob/da12d7ac8a74a7497cd0d640ed3c2a301a551ced/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go#L34) do this conversion, while taking both `MatchLabels` and `MatchExpressions` into consideration.
2. And then the scheduler checks if one Pod matches with the label selector via `(labels.Selector).Matches()` ([ref](https://github.com/kubernetes/kubernetes/blob/da12d7ac8a74a7497cd0d640ed3c2a301a551ced/pkg/scheduler/framework/types.go#L282))

Does it answer your question?